### PR TITLE
#162170712 Create users profiles

### PR DESCRIPTION
**Pull request by oluwajuwon**
#### What does this PR do?
This PR handles retrieving a logged in or signed up user's profile, editing a user's profile and getting other user's profile
#### Description of Task to be completed?
This task has the following endpoints working

- GET /api/user/
To retrieve a current user's profile/details
- PUT /api/user/
To update a current user's profile/details
- GET /api/profiles/:username
To get other user's profile

#### How should this be manually tested?
- Clone this repo
- In your terminal cd into the folder it creates
- Run **npm install** to install all the dependencies that will be used by the project
- Check the **.env.sample** to create a **.env** file in your root folder
- Run **npm run dev:start** to start the app
- Signup through **localhost:3000/api/users/signup** or Login through **localhost:3000/api/users/login** respectively using postman as a client
- Send a GET request to **localhost:3000/api/user** to get your profile details
- Send a PUT request to **localhost:3000/api/user** with the required fields (username, email, bio, image) filled to update your profile details
- Send a GET request to **localhost:3000/api/profiles/(enter username)** to retrieve a user's profile
- Run **npm test** to test the app
- View the Swagger UI on **localhost:3000/api/docs**
#### Any background context you want to provide?
N/A
#### What are the relevant pivotal tracker stories?
[#162170712](https://www.pivotaltracker.com/story/show/162170712)
#### Screenshots (if appropriate)
screen shot 2018-12-04 at 4 19 54 pm
screen shot 2018-12-04 at 4 20 15 pm
screen shot 2018-12-04 at 4 20 18 pm

#### Questions:

### CODE REVIEW

File: src/controllers/profile.js [hunk 1]
```diff
@@ -0,0 +1,140 @@
...
+        return response.status(200).json({
+          status: 'Success',
+          message: 'Retrieved user successfully',
+          currentUser
+        });
+      }
+    } catch (error) {
+      response.status(500).json({
+        status: 'Fail',
+        message: 'Could not retrieve user at this moment',
```
Comment (frostyblok) [inline src/controllers/profile.js]: Nice one Juwon, I would like to know why you were sending back a custom error message and not the one from the callback. You could have something like 'error: error.message'.

File: src/controllers/profile.js [hunk 2]
```diff
@@ -0,0 +1,140 @@
...
+      if (updatedUser && !updatedUser.error) {
+        return response.status(200).json({
+          status: 'Success',
+          message: 'Updated profile successfully'
+        });
+      }
+    } catch (error) {
+      response.status(500).json({
+        status: 'Fail',
+        message: 'Can not update profile at this moment',
```
Comment (frostyblok) [inline src/controllers/profile.js]: Same as above...

File: src/controllers/users.js [hunk 3]
```diff
@@ -40,51 +40,47 @@ class userHandler {
...
+      });
     } catch (error) {
-      return response.status(500)
-        .json({
-          status: 'Fail',
-          error
-        });
+      response.status(500).json({
+        status: 'Fail',
+        error: error.message
```
Comment (frostyblok) [inline src/controllers/users.js]: Good job here!

File: src/docs/swagger.js [hunk 4]
```diff
@@ -26,11 +26,15 @@ export default {
   tags: [
     {
       name: 'users',
-      description: 'The users of Author\'s Haven'
+      description: "The users of Author's Haven"
     },
     {
       name: 'articles',
-      description: 'The articles created by Author\'s Haven users'
+      description: "The articles created by Author's Haven users"
```
Comment (frostyblok) [inline src/docs/swagger.js]: Juwon, I think you should change this to a single quote.

**Review by frostyblok — CHANGES_REQUESTED** [review]
This is a beautiful job, Juwon. I like I how you were able to reduce the complexity of your code block, I just want you to make some changes as regards the error message of the callback you were not returning. Good job once again.

File: src/controllers/profile.js (hunk 1 above)
Comment (oluwajuwon) [reply src/controllers/profile.js]: Nice observation, I felt the user might not be able to understand the error message It throws, but from the small research I just did I've found it to be quite descriptive. Thanks, I'll update it now

File: src/controllers/profile.js (hunk 2 above)
Comment (oluwajuwon) [reply src/controllers/profile.js]: Ok, done

File: src/controllers/users.js (hunk 3 above)
Comment (oluwajuwon) [reply src/controllers/users.js]: Thank you

File: src/docs/swagger.js (hunk 4 above)
Comment (oluwajuwon) [reply src/docs/swagger.js]: Thanks for the observation, that was a mistake on my path. I'll fix that right away

**Review by marcdomain — COMMENTED** [review]
The authentication endpoints are **/api/users/signup** & **/api/users/login** not **/api/signup** & **/api/login**

**Review by Johnsonojo — COMMENTED** [review]
Looks good to me, except for the changes that have been pointed out for you to implement.

**Review by marcdomain — APPROVED** [review]
LGTM!

File: src/controllers/users.js (hunk 3 above)
Comment (frostyblok) [reply src/controllers/users.js]: Good job, Juwon.

**Review by frostyblok — APPROVED** [review]

**Pull request merged**